Amplitude should record the user's balance (amount of gems)

Our  Amplitude analytics data should include the user's number of gems (the user's `balance` object) to help us track payment / purchase problems.

Ideally it should report it as something like this `Balance   0.75 (3 Gems)` to avoid us making mistakes with the mental arithmetic - I know it's not hard, but it's even easier for the computer. ;)

If for any reason the balance doesn't exist or isn't a finite number, the code should report that (I have no idea if that's possible with our current code base or any future code, but it wouldn't be good to report a balance of 0 if there was actually some bug that prevented the balance existing at all).

**EDIT 2016-12-27: The work remaining to do is:**

- Leave the existing "balance" property as it is now.
- Create a new property called "balance gem amount" or similar (spaces are optional) which works in the same way as the "balance" property but displays the balance divided by 4. I want the word "balance" (lower case) to be at the start of the property name so that in Amplitude it appears immediately under the existing "balance" field (sorting is alphabetical there).

Hi a question again.
when I try to run test/server_side/analytics.test.js following error pops up.

module.js:328
    throw err;
    ^

Error: Cannot find module '../../website/server/libs/api-v2/analytics'

--- a/test/api/v3/unit/libs/analyticsService.test.js
+++ b/test/api/v3/unit/libs/analyticsService.test.js
@@ -277,6 +277,7 @@ describe('analyticsService', () => {
           dailys: [{_id: 'daily'}],
           todos: [{_id: 'todo'}],
           rewards: [{_id: 'reward'}],
+          balance: 12,
         };
 
         data.user = user;
@@ -300,6 +301,7 @@ describe('analyticsService', () => {
                 },
                 contributorLevel: 1,
                 subscription: 'foo-plan',
+                balance: 12,
               },
             });
           });

--- a/test/client-old/spec/services/analyticsServicesSpec.js
+++ b/test/client-old/spec/services/analyticsServicesSpec.js
@@ -193,6 +193,7 @@ describe('Analytics Service', function () {
           todos: 1,
           rewards: 1
         };
+        expectedProperties.balance = 12;
 
         beforeEach(function() {
           user._id = 'unique-user-id';
@@ -207,6 +208,7 @@ describe('Analytics Service', function () {
           user.dailys = [{_id: 'daily'}];
           user.todos = [{_id: 'todo'}];
           user.rewards = [{_id: 'reward'}];
+          user.balance = 12;
 
           analytics.updateUser(properties);
           clock.tick();
@@ -240,7 +242,8 @@ describe('Analytics Service', function () {
             dailys: 1,
             habits: 1,
             rewards: 1
-          }
+          },
+          balance: 12
         };
 
         beforeEach(function() {
@@ -258,6 +261,7 @@ describe('Analytics Service', function () {
           user.dailys = [{_id: 'daily'}];
           user.todos = [{_id: 'todo'}];
           user.rewards = [{_id: 'reward'}];
+          user.balance = 12;
 
           analytics.updateUser();
           clock.tick();

--- a/website/client-old/js/services/analyticsServices.js
+++ b/website/client-old/js/services/analyticsServices.js
@@ -127,6 +127,11 @@
       properties.Level = user.stats.lvl;
       properties.Mana = Math.floor(user.stats.mp);
     }
+
+    if (user.balance) {
+      properties.balance = user.balance;
+    }
+
     properties.tutorialComplete = user.flags && user.flags.tour && user.flags.tour.intro === -2;
     if (user.habits && user.dailys && user.todos && user.rewards) {
       properties["Number Of Tasks"] = {

--- a/website/client-old/js/services/analyticsServices.js
+++ b/website/client-old/js/services/analyticsServices.js
@@ -128,9 +128,7 @@
       properties.Mana = Math.floor(user.stats.mp);
     }
 
-    if (user.balance) {
-      properties.balance = user.balance;
-    }
+    properties.balance = user.balance;
 
     properties.tutorialComplete = user.flags && user.flags.tour && user.flags.tour.intro === -2;
     if (user.habits && user.dailys && user.todos && user.rewards) {

--- a/website/server/libs/analyticsService.js
+++ b/website/server/libs/analyticsService.js
@@ -66,6 +66,10 @@ let _formatUserData = (user) => {
     properties.Mana = Math.floor(user.stats.mp);
   }
 
+  if (user.balance) {
+    properties.balance = user.balance;
+  }
+
   properties.tutorialComplete = user.flags && user.flags.tour && user.flags.tour.intro === -2;
 
   if (user.habits && user.dailys && user.todos && user.rewards) {

--- a/website/server/libs/analyticsService.js
+++ b/website/server/libs/analyticsService.js
@@ -66,9 +66,7 @@ let _formatUserData = (user) => {
     properties.Mana = Math.floor(user.stats.mp);
   }
 
-  if (user.balance) {
-    properties.balance = user.balance;
-  }
+  properties.balance = user.balance;
 
   properties.tutorialComplete = user.flags && user.flags.tour && user.flags.tour.intro === -2;
 

